iree/compiler/Dialect/Flow/IR/FlowOps.cpp

```diff
@@ -580,101 +731,71 @@ static void printDispatchWorkgroupBody(OpAsmPrinter &p, Operation *op,
-  ::llvm::ArrayRef<::mlir::Type> resultsTypes;
-  std::unique_ptr<::mlir::Region> bodyRegion =
-      std::make_unique<::mlir::Region>();
-  if (parser.parseLSquare()) return ::mlir::failure();
-
-  workgroup_countOperandsLoc = parser.getCurrentLocation();
-  if (parser.parseOperandList(workgroup_countOperands))
-    return ::mlir::failure();
-  if (parser.parseRSquare()) return ::mlir::failure();
-  if (parser.parseLParen()) return ::mlir::failure();
-
-  operandsOperandsLoc = parser.getCurrentLocation();
-  if (parser.parseOperandList(operandsOperands)) return ::mlir::failure();
-  if (parser.parseRParen()) return ::mlir::failure();
-  if (parser.parseColon()) return ::mlir::failure();
-
-  ::mlir::FunctionType operands__results_functionType;
-  if (parser.parseType(operands__results_functionType))
-    return ::mlir::failure();
-  operandsTypes = operands__results_functionType.getInputs();
-  resultsTypes = operands__results_functionType.getResults();
-  if (parser.parseOptionalAttrDictWithKeyword(result->attributes))
-    return ::mlir::failure();
-  if (parser.parseEqual()) return ::mlir::failure();
-  {
-    if (parseDispatchWorkgroupBody(parser, operandsTypes, resultsTypes,
-                                   *bodyRegion))
-      return ::mlir::failure();
-  }
-  ::mlir::Type odsBuildableType0 = parser.getBuilder().getIndexType();
-  result->addTypes(resultsTypes);
-  if (parser.resolveOperands(workgroup_countOperands, odsBuildableType0,
-                             result->operands))
-    return ::mlir::failure();
-  if (parser.resolveOperands(operandsOperands, operandsTypes,
-                             operandsOperandsLoc, result->operands))
-    return ::mlir::failure();
-  result->addRegion(std::move(bodyRegion));
-  result->addAttribute(
-      "operand_segment_sizes",
-      parser.getBuilder().getI32VectorAttr(
-          {static_cast<int32_t>(workgroup_countOperands.size()),
-           static_cast<int32_t>(operandsOperands.size())}));
-  return ::mlir::success();
-}
-
-static void printDispatchWorkgroupsOp(::mlir::OpAsmPrinter &p,
-                                      DispatchWorkgroupsOp &op) {
-  p << "flow.dispatch.workgroups";
-  p << "[";
-  p << op.workgroup_count();
-  p << "]";
-  p << ' ' << "(";
-  p << op.operands();
-  p << ")";
-  p << ' ' << ":";
-  p << ' ';
-  p.printFunctionalType(op.operands().getTypes(), op.results().getTypes());
-  p.printOptionalAttrDictWithKeyword(op->getAttrs(), /*elidedAttrs=*/{
-                                         "operand_segment_sizes",
-                                     });
-  p << ' ' << "=";
-  p << ' ';
-  printDispatchWorkgroupBody(p, op, op.operands().getTypes(),
-                             op.results().getTypes(), op.body());
-}
-
 static LogicalResult verifyDispatchWorkgroupsOp(DispatchWorkgroupsOp op) {
   if (op.workgroup_count().empty()) {
     return op.emitOpError() << "at least one workgroup dimension is required";
```

Comment:
this came to me last night: one ineresting verification to add is that operand_dims().elngth() is equal to the number of `?`'s? This can help with the situation where a fold (incorrectly) changes the result type of its op, perhaps refining the type to have fewer `?` (we have some in linalg I believe), which can reduce the `?`'s of an operand to this function. That should make it a nicer error instead of a miscompile / segfault.

Reply:
good thinking! I'll make a helper and use it on all of these ops.

Reply:
and it found a bug in a test that I didn't update!
```
error: 'flow.tensor.update' op value set has 2 dynamic dimensions but only 0 dimension values are attached
```